scipy/stats/stats.py

```diff
@@ -1888,24 +1894,71 @@ def scoreatpercentile(a, per, limit=(), interpolation_method='fraction',
+        raise ValueError("`per` must be 1d.")
+
+    # patch for numpy/numpy#18158. nan
+    # If zero size, only returned shape matters.
     if a.size == 0:
-        # empty array, return nan(s) with shape matching `per`
-        if np.isscalar(per):
-            return np.nan
-        else:
-            return np.full(np.asarray(per).shape, np.nan, dtype=np.float64)
+        with warnings.catch_warnings() as _, \
+                np.errstate(invalid='ignore') as _:
+            warnings.filterwarnings("ignore", "Mean of empty slice",
+                                    RuntimeWarning)
+            res = np.mean(a, axis=axis)
+            if np.ndim(per):
+                return np.broadcast_to(res, np.shape(per) + res.shape)
+            else:
+                return res
 
-    if limit:
-        a = a[(limit[0] <= a) & (a <= limit[1])]
+    contains_nan, nan_policy = _contains_nan(a, nan_policy)
 
-    sorted_ = np.sort(a, axis=axis)
-    if axis is None:
-        axis = 0
+    if nan_policy == 'omit':
+        percentile_fun = np.nanpercentile
+    elif nan_policy == 'propagate':
+        percentile_fun = np.percentile
 
-    return _compute_qth_percentile(sorted_, per, interpolation_method, axis)
+    if interpolation_method == 'fraction':
+        interpolation = 'linear'
+    else:
+        interpolation = interpolation_method
+
+    if limit:  # everything is simple.... unless there are limits
+        with np.errstate(invalid='ignore'):
+            if axis is None:
+                # the array gets flattened, so we can just keep all
+                # values within limits and all nans
+                a = a[(limit[0] <= a) & (a <= limit[1])
+                      | np.isnan(a)]
+            elif nan_policy == 'omit' or not(contains_nan):
+                # either way, it is safe to use `nanpercentile`, replacing
+                # values beyond limits with nans
+                a = a.astype(float)  # to store nans
+                a[(limit[0] > a) | (a > limit[1])] = np.nan
+                percentile_fun = np.nanpercentile
+            elif nan_policy == 'propagate':
+                # we have to distinguish between nans in the data (which cause
+                # nans in output) and nans that come from values beyond limits
+                a = a.astype(float)  # to store nans
+                output_nan = np.any(np.isnan(a), axis=axis)
+                a[(limit[0] > a) | (a > limit[1])] = np.nan
+                with warnings.catch_warnings():
+                    warnings.filterwarnings("ignore",
+                                            "All-NaN slice encountered",
+                                            RuntimeWarning)
+                    res = np.nanpercentile(a=a, q=per, axis=axis,
+                                           interpolation=interpolation)
+                res[..., output_nan] = np.nan
+                return res
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore",
```

Comment:
FWIW, I do see an example of the NumPy `suppress_warnings` outside of a test in:
`scipy/stats/_binned_statistic.py`

Some residual exceptions to the rule are explicitly exempted in `test_warning_calls_filters` as you likely noticed, but I don't see the above module mentioned there. Not sure that's justification to use it, but just observing..